source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -38,165 +38,163 @@ namespace spvtools {
 namespace opt {
 
 namespace {
-// The core operate function to get spec operation result with the given opcode
-// and operands in 32 bit words. This function returns the result in 32 bit
-// word.
-uint32_t OperateWords(SpvOp opcode,
-                      const std::vector<uint32_t>& operand_words) {
+// Returns the single-word result from performing the given unary operation on
+// the operand value which is passed in as a 32-bit word.
+uint32_t UnaryOperate(SpvOp opcode, uint32_t operand) {
   switch (opcode) {
     // Arthimetics
     case SpvOp::SpvOpSNegate:
       // TODO(qining): refer to grammar for such asserts.
-      assert(operand_words.size() == 1u);
-      return -static_cast<int32_t>(operand_words.front());
+      return -static_cast<int32_t>(operand);
     case SpvOp::SpvOpNot:
-      assert(operand_words.size() == 1u);
-      return ~operand_words.front();
+      return ~operand;
+    case SpvOp::SpvOpLogicalNot:
+      return !static_cast<bool>(operand);
+    default:
+      assert(false &&
+             "Unsupported unary operation for OpSpecConstantOp instruction");
+      return 0u;
+  }
+}
+
+// Returns the single-word result from performing the given binary operation on
+// the operand values which are passed in as two 32-bit word.
+uint32_t BinaryOperate(SpvOp opcode, uint32_t a, uint32_t b) {
+  switch (opcode) {
+    // Arthimetics
+    case SpvOp::SpvOpSNegate:
+    // TODO(qining): refer to grammar for such asserts.
     case SpvOp::SpvOpIAdd:
-      assert(operand_words.size() == 2u);
-      return operand_words.front() + operand_words.back();
+      return a + b;
     case SpvOp::SpvOpISub:
-      assert(operand_words.size() == 2u);
-      return operand_words.front() - operand_words.back();
+      return a - b;
     case SpvOp::SpvOpIMul:
-      assert(operand_words.size() == 2u);
-      return operand_words.front() * operand_words.back();
+      return a * b;
     case SpvOp::SpvOpUDiv:
-      assert(operand_words.size() == 2u);
-      assert(operand_words.back() != 0);
-      return operand_words.front() / operand_words.back();
+      assert(b != 0);
+      return a / b;
     case SpvOp::SpvOpSDiv:
-      assert(operand_words.size() == 2u);
-      assert(operand_words.back() != 0u);
-      return (static_cast<int32_t>(operand_words.front())) /
-             (static_cast<int32_t>(operand_words.back()));
+      assert(b != 0u);
+      return (static_cast<int32_t>(a)) / (static_cast<int32_t>(b));
     case SpvOp::SpvOpSRem: {
-      assert(operand_words.size() == 2u);
-      assert(operand_words.back() != 0u);
-      int32_t a = static_cast<int32_t>(operand_words.front());
-      int32_t b = static_cast<int32_t>(operand_words.back());
-      return a % b;
+      // The sign of non-zero result comes from the first operand: a
```

Comment:
Right, but you should explain why % works here.
You're relying on C++11 rule for integer division operator, since the quotient is rounded toward zero.  That's what makes non-zero modulus have the sign of operand 1.
(I had to go through the 4 possible cases of signs of operands 1 and 2 to verify this.)

Needs comment!